tests/command/helpers/getschemavalidranges.js

```diff
@@ -21,21 +22,52 @@ describe( 'getSchemaValidRanges', () => {
 		schema.registerItem( 'h1', '$block' );
 		schema.registerItem( 'img', '$inline' );
 
-		// Allow bold on p
 		schema.allow( { name: '$text', attributes: 'bold', inside: 'p' } );
 		schema.allow( { name: 'p', attributes: 'bold', inside: '$root' } );
 
 		setData( document, '<p>foo<img />bar</p>' );
 		ranges = [ Range.createOn( root.getChild( 0 ) ) ];
 	} );
 
-	it( 'should return unmodified ranges when attribute is allowed on each element', () => {
-		// Allow bold on img
+	it( 'should return unmodified ranges when attribute is allowed on each item (v1 – text is not allowed in img)', () => {
 		schema.allow( { name: 'img', attributes: 'bold', inside: 'p' } );
 
 		expect( getSchemaValidRanges( attribute, ranges, schema ) ).to.deep.equal( ranges );
 	} );
 
+	it( 'should return unmodified ranges when attribute is allowed on each item (v2 – text is allowed in img)', () => {
+		schema.allow( { name: 'img', attributes: 'bold', inside: 'p' } );
+		schema.allow( { name: '$text', inside: 'img' } );
+
+		expect( getSchemaValidRanges( attribute, ranges, schema ) ).to.deep.equal( ranges );
+	} );
+
+	it( 'should return two ranges when attribute is not allowed in one item', () => {
+		schema.allow( { name: 'img', attributes: 'bold', inside: 'p' } );
+		schema.allow( { name: '$text', inside: 'img' } );
+
+		setData( document, '<p>foo<img>xxx</img>bar</p>' );
+
+		const validRanges = getSchemaValidRanges( attribute, ranges, schema );
+		const sel = new Selection();
+		sel.setRanges( validRanges );
+
+		expect( stringify( root, sel ) ).to.equal( '[<p>foo<img>]xxx[</img>bar</p>]' );
```

Comment:
I think that nothing's gonna happen because such a range has two empty flat ranges. And the first range contains `[bar]` and `[]</img>`, so I presume it will only affect the text. Still, we'd need to check this...


EDIT: I started checking it and I realised that attr op can be applied to non-flat ranges. Later on the writer uses `range.getItems()`, though, which ignores element ends, so it seems that I was right anyway.